File: crates/sobria-core/src/preferences.rs
//! Personas et identifiants stables des modules — voir ADR-0010 et
//! CDC v1.3 §3 (personas) + §4 (25 modules).
//!
//! Ces types sont **fermés** : ajouter un persona ou un module nécessite
//! un bump de version de schéma (chantier dédié + ADR mini-incrément).
//!
//! Ils vivent dans `sobria-core` parce qu'ils sont consommés à la fois par
//! `sobria-app` (préférences IPC) et par tout futur frontend tiers
//! (extension navigateur, CLI, etc.).

use schemars::JsonSchema;
use serde::{Deserialize, Serialize};

/// Les cinq personas figés en v1.3 du cahier des charges.
///
/// Un utilisateur choisit un et un seul persona à la fois ; il peut
/// néanmoins personnaliser librement ses modules au-delà du bundle
/// par défaut. Voir ADR-0010 §"Principes" pour la justification.
#[derive(Debug, Clone, Copy, Serialize, Deserialize, JsonSchema, PartialEq, Eq, Hash)]
#[serde(rename_all = "snake_case")]
pub enum Persona {
    /// Étudiant·e ou simple curieux·se — découverte, apprentissage, usage perso.
    Student,
    /// Professionnel·le tech (dev, ML eng, intégrateur).
    ProTech,
    /// Entreprise — DSI, RSE, scope 3 IA, reporting CSRD.
    Enterprise,
    /// Collectivité ou service public — territorial, marchés publics.
    PublicSector,
    /// Chercheur·se ou journaliste — reproductibilité, comparaisons sourcées.
    Researcher,
}

impl Persona {
    /// Itère sur les 5 personas dans l'ordre canonique du CDC §3.
    #[must_use]
    pub fn all() -> &'static [Persona] {
        &[
            Persona::Student,
            Persona::ProTech,
            Persona::Enterprise,
            Persona::PublicSector,
            Persona::Researcher,
        ]
    }

    /// Bundle de modules pré-cochés au moment où l'utilisateur sélectionne
    /// ce persona dans l'onboarding. L'utilisateur peut ensuite ajouter
    /// ou retirer librement n'importe quel module.
    ///
    /// **Périmètre v1.0 figé** (cf. ADR-0011 — réduction périmètre) :
    /// seuls **13 modules** sont en cible v1.0 — M1, M3, M7, M8, M9, M12,
    /// M13, M14, M15, M17, M20, M22, M25. Les autres `ModuleId` restent
    /// dans l'enum (forward compat, activables manuellement) mais ne
    /// figurent dans aucun bundle par défaut.
    ///
    /// **Invariants garantis par les tests** :
    /// - le bundle est non vide,
    /// - `ModuleId::M1` (Estimer) est présent dans tous les bundles,
    /// - aucun doublon,
    /// - aucune référence à un identifiant `M4` (réservé en v1.3),
    /// - tous les modules d'un bundle sont dans le set v1.0 (13 IDs).
    #[must_use]
    pub fn default_modules(self) -> Vec<ModuleId> {
        use ModuleId::{
            M1, M12, M13, M14, M15, M17, M20, M22, M25, M3, M7, M8, M9,
        };
        match self {
            // Étudiant / Curieux : apprendre, comprendre, suivre, fixer.
            Persona::Student => vec![M1, M8, M13, M14, M15, M25],
            // Pro tech : estimer, comparer, journal, refs, méthodo.
            Persona::ProTech => vec![M1, M3, M7, M8, M9, M13, M14],
            // Entreprise : toute la chaîne compliance + dashboard + budget.
            Persona::Enterprise => {
                vec![M1, M7, M12, M14, M15, M17, M20, M22, M25]
            },
            // Collectivité : focus territoire FR + reporting.
            Persona::PublicSector => vec![M1, M8, M12, M14, M17, M20, M22],
            // Chercheur / Journaliste : méthodologie + reproductibilité.
            Persona::Researcher => vec![M1, M3, M7, M8, M9, M14, M17],
        }
    }

    /// Liste des modules en cible v1.0 (13 IDs). Voir ADR-0011.
    #[must_use]
    pub fn v1_0_modules() -> &'static [ModuleId] {
        use ModuleId::{M1, M12, M13, M14, M15, M17, M20, M22, M25, M3, M7, M8, M9};
        &[M1, M3, M7, M8, M9, M12, M13, M14, M15, M17, M20, M22, M25]
    }
}

/// Identifiants stables des 24 modules visibles en v1.3 du CDC.
///
/// L'identifiant `M4` est volontairement **omis** : il était utilisé en
/// v1.2 et a été retiré dans v1.3 pour éviter la confusion. Tout code
/// produisant ou consommant la string `"m4"` est invalide.
///
/// L'ordre déclaratif ci-dessous correspond à l'ordre par défaut dans
/// le rail UI (voir `web/src/routes/+layout.svelte`).
#[derive(Debug, Clone, Copy, Serialize, Deserialize, JsonSchema, PartialEq, Eq, Hash)]
#[serde(rename_all = "lowercase")]
pub enum ModuleId {
    /// M1 — Estimer un prompt (cœur produit, dans tous les bundles).
    M1,
    /// M2 — Workbench multi-prompts.
    M2,
    /// M3 — Comparer modèles (benchmark côte-à-côte).
    M3,
    // M4 — réservé / retiré v1.3.
    /// M5 — Exporter rapport (PDF/CSV/JSON).
    M5,
    /// M6 — Géoloc datacenter unitaire.
    M6,
    /// M7 — Journal d'audit (ledger chaîné SHA-256).
    M7,
    /// M8 — Méthodologie interactive.
    M8,
    /// M9 — Référentiel modèles (catalogue 25+).
    M9,
    /// M10 — Importer batch (CSV/JSON multi-prompts).
    M10,
    /// M11 — Extension navigateur (capture vie réelle).
    M11,
    /// M12 — Datacenters Europe (carte OpenStreetMap + drill-down).
    M12,
    /// M13 — Simulateur « Et si...? » (7 leviers temps réel).
    M13,
    /// M14 — À propos / Crédits / Licences.
    M14,
    /// M15 — Tableau de bord personnel (jour/semaine/mois).
    M15,
    /// M16 — Forecaster 12 mois (bande d'incertitude × sliders).
    M16,
    /// M17 — Empreinte projet (reproductibilité scientifique).
    M17,
    /// M18 — Batch CSV → rapport agrégé.
    M18,
    /// M19 — Équipe / multi-utilisateurs (RBAC léger).
    M19,
    /// M20 — Territoire FR (RTE IRIS, Sankey énergétique).
    M20,
    /// M21 — Alertes & seuils (notifs locales).
    M21,
    /// M22 — Rapport CSRD / AGEC / AFNOR SPEC 2314 (PDF signé + PROV-O).
    M22,
    /// M23 — Marchés publics IA frugale (cahiers des charges types).
    M23,
    /// M24 — Apprendre (mini-cours, best practices).
    M24,
    /// M25 — Objectifs & habitudes (eco-budget personnel).
    M25,
}

impl ModuleId {
    /// Énumère les 24 modules dans l'ordre canonique du rail UI.
    #[must_use]
    pub fn all() -> &'static [ModuleId] {
        use ModuleId::{
            M1, M10, M11, M12, M13, M14, M15, M16, M17, M18, M19, M2, M20, M21, M22, M23, M24,
            M25, M3, M5, M6, M7, M8, M9,
        };
        &[
            M1, M2, M3, M5, M6, M7, M8, M9, M10, M11, M12, M13, M14, M15, M16, M17, M18, M19,
            M20, M21, M22, M23, M24, M25,
        ]
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::collections::HashSet;

    #[test]
    fn all_personas_have_non_empty_bundles() {
        for p in Persona::all() {
            assert!(
                !p.default_modules().is_empty(),
                "persona {p:?} a un bundle vide"
            );
        }
    }

    #[test]
    fn m1_is_in_every_bundle() {
        for p in Persona::all() {
            assert!(
                p.default_modules().contains(&ModuleId::M1),
                "persona {p:?} ne contient pas M1 (point fixe)"
            );
        }
    }

    #[test]
    fn bundles_have_no_duplicates() {
        for p in Persona::all() {
            let bundle = p.default_modules();
            let unique: HashSet<_> = bundle.iter().collect();
            assert_eq!(
                bundle.len(),
                unique.len(),
                "doublons dans le bundle {p:?} : {bundle:?}"
            );
        }
    }

    #[test]
    fn all_modules_excludes_m4() {
        let json_ids: Vec<String> = ModuleId::all()
            .iter()
            .map(|m| serde_json::to_string(m).unwrap())
            .collect();
        for id in &json_ids {
            assert_ne!(id, "\"m4\"", "M4 doit être réservé / absent");
        }
        // Pour mémoire, la liste doit aussi contenir m1 et m25
        assert!(json_ids.contains(&"\"m1\"".to_string()));
        assert!(json_ids.contains(&"\"m25\"".to_string()));
    }

    #[test]
    fn all_modules_count_is_24() {
        // 25 - 1 (M4 réservé) = 24 modules visibles.
        assert_eq!(ModuleId::all().len(), 24);
    }

    #[test]
    fn persona_serde_round_trip() {
        for p in Persona::all() {
            let json = serde_json::to_string(p).unwrap();
            let back: Persona = serde_json::from_str(&json).unwrap();
            assert_eq!(back, *p);
        }
    }

    #[test]
    fn persona_serializes_snake_case() {
        assert_eq!(
            serde_json::to_string(&Persona::ProTech).unwrap(),
            "\"pro_tech\""
        );
        assert_eq!(
            serde_json::to_string(&Persona::PublicSector).unwrap(),
            "\"public_sector\""
        );
    }

    #[test]
    fn module_id_serde_round_trip() {
        for m in ModuleId::all() {
            let json = serde_json::to_string(m).unwrap();
            let back: ModuleId = serde_json::from_str(&json).unwrap();
            assert_eq!(back, *m);
        }
    }

    #[test]
    fn module_id_serializes_lowercase() {
        assert_eq!(serde_json::to_string(&ModuleId::M1).unwrap(), "\"m1\"");
        assert_eq!(serde_json::to_string(&ModuleId::M25).unwrap(), "\"m25\"");
    }

    #[test]
    fn module_id_deserialize_m4_fails() {
        let result: Result<ModuleId, _> = serde_json::from_str("\"m4\"");
        assert!(result.is_err(), "m4 doit être rejeté à la désérialisation");
    }

    #[test]
    fn deserialize_unknown_persona_fails() {
        let result: Result<Persona, _> = serde_json::from_str("\"hacker\"");
        assert!(result.is_err());
    }

    #[test]
    fn each_persona_bundle_modules_are_in_all_modules() {
        let all: HashSet<ModuleId> = ModuleId::all().iter().copied().collect();
        for p in Persona::all() {
            for m in p.default_modules() {
                assert!(
                    all.contains(&m),
                    "module {m:?} du bundle {p:?} pas dans ModuleId::all()"
                );
            }
        }
    }

    #[test]
    fn v1_0_modules_returns_13() {
        // ADR-0011 : périmètre v1.0 figé à 13 modules essentiels.
        assert_eq!(Persona::v1_0_modules().len(), 13);
    }

    #[test]
    fn v1_0_modules_includes_pivot_modules() {
        let v1 = Persona::v1_0_modules();
        // Modules clés du pitch
        for required in [
            ModuleId::M1,  // Estimer
            ModuleId::M7,  // Journal d'audit
            ModuleId::M20, // Territoire FR (différenciateur)
            ModuleId::M22, // Rapport CSRD
            ModuleId::M17, // Empreinte projet (datasheet Gebru)
        ] {
            assert!(
                v1.contains(&required),
                "module pivot {required:?} absent du périmètre v1.0"
            );
        }
    }

    #[test]
    fn v1_0_modules_excludes_deferred_modules() {
        let v1: HashSet<ModuleId> = Persona::v1_0_modules().iter().copied().collect();
        // Modules différés v1.1+ (cf. ADR-0011)
        for deferred in [
            ModuleId::M2,  // Workbench
            ModuleId::M5,  // Rapports génériques
            ModuleId::M6,  // Géoloc unitaire
            ModuleId::M10, // Import logs
            ModuleId::M11, // Extension navigateur
            ModuleId::M16, // Forecaster (backend prêt mais UI v1.1)
            ModuleId::M18, // Batch CSV (idem)
            ModuleId::M19, // Équipe
            ModuleId::M21, // Alertes
            ModuleId::M23, // Marchés publics
            ModuleId::M24, // Apprendre
        ] {
            assert!(
                !v1.contains(&deferred),
                "module {deferred:?} ne doit pas être dans le périmètre v1.0"
            );
        }
    }

    #[test]
    fn all_persona_bundles_only_reference_v1_0_modules() {
        let v1: HashSet<ModuleId> = Persona::v1_0_modules().iter().copied().collect();
        for p in Persona::all() {
            for m in p.default_modules() {
                assert!(
                    v1.contains(&m),
                    "persona {p:?} référence {m:?} hors périmètre v1.0"
                );
            }
        }
    }
}

File: web/src/lib/preferences.ts
// Store de préférences utilisateur (C10 — ADR-0010).
//
// Source de vérité : `crates/sobria-core/src/preferences.rs` (Persona +
// ModuleId) et `crates/sobria-app/src/dto.rs` (AppPreferencesDto). Toute
// dérive entre ces deux fichiers TS et Rust doit casser un test e2e ou
// un test Rust serde (cf. `each_persona_bundle_modules_are_in_all_modules`
// dans sobria-core).
//
// Optimistic update + rollback : `savePreferences` met à jour le store
// localement avant l'IPC. Si l'IPC échoue, on restaure l'état précédent
// et on rethrow — l'UI affiche alors la bannière d'erreur.

import { writable, get } from 'svelte/store';
import {
  getAppPreferences,
  setAppPreferences,
  type AppPreferencesDto,
  type ModuleId,
  type Persona
} from './api';

export type { ModuleId, Persona } from './api';
export type AppPreferences = AppPreferencesDto;

// État du store : DTO + flag `loaded` pour distinguer "pas encore chargé"
// (rail unfiltered) de "chargé avec 0 module" (rail vide légitime).
export interface PreferencesState extends AppPreferencesDto {
  loaded: boolean;
}

const INITIAL: PreferencesState = {
  persona: null,
  enabled_modules: [],
  onboarded: false,
  lang: 'fr',
  default_method: 'afnor_sobria',
  also_show_methods: [],
  default_datacenter_id: undefined,
  loaded: false
};

export const preferences = writable<PreferencesState>(INITIAL);

/**
 * Charge les préférences depuis l'IPC et met à jour le store.
 * Doit être appelé une fois au boot (cf. `+layout.svelte`).
 * Throw si IPC indisponible — l'appelant gère.
 */
export async function loadPreferences(): Promise<void> {
  const p = await getAppPreferences();
  preferences.set({ ...p, loaded: true });
}

/**
 * Persiste les préférences via IPC. Optimistic update + rollback.
 */
export async function savePreferences(p: AppPreferencesDto): Promise<void> {
  const prev = get(preferences);
  preferences.set({ ...p, loaded: true });
  try {
    await setAppPreferences(p);
  } catch (e) {
    preferences.set(prev);
    throw e;
  }
}

/** Catalogue ordonné des 24 personas (mirror `sobria_core::Persona::all`). */
export const ALL_PERSONAS: readonly Persona[] = [
  'student',
  'pro_tech',
  'enterprise',
  'public_sector',
  'researcher'
] as const;

/** Catalogue ordonné des 24 modules (mirror `sobria_core::ModuleId::all`). */
export const ALL_MODULES: readonly ModuleId[] = [
  'm1',
  'm2',
  'm3',
  'm5',
  'm6',
  'm7',
  'm8',
  'm9',
  'm10',
  'm11',
  'm12',
  'm13',
  'm14',
  'm15',
  'm16',
  'm17',
  'm18',
  'm19',
  'm20',
  'm21',
  'm22',
  'm23',
  'm24',
  'm25'
] as const;

/**
 * Bundle pré-coché par persona. Mirror exact de
 * `sobria_core::Persona::default_modules` — tout drift fait tomber les
 * tests Rust + le test e2e Playwright qui valide la cohérence.
 *
 * **C32.1** : alignement strict sur le canon Rust + retrait de M14 (page
 * « À propos » accessible via menu, pas un module fonctionnel à pré-cocher).
 */
export function defaultModulesFor(persona: Persona | null): ModuleId[] {
  switch (persona) {
    case 'student':
      return ['m1', 'm8', 'm13', 'm15', 'm25'];
    case 'pro_tech':
      return ['m1', 'm3', 'm7', 'm8', 'm9', 'm13'];
    case 'enterprise':
      return ['m1', 'm7', 'm12', 'm15', 'm17', 'm20', 'm22', 'm25'];
    case 'public_sector':
      return ['m1', 'm8', 'm12', 'm17', 'm20', 'm22'];
    case 'researcher':
      return ['m1', 'm3', 'm7', 'm8', 'm9', 'm17'];
    case null:
      return [];
  }
}

// ─── Catalogue persona ───────────────────────────────────────────────────

export interface PersonaInfo {
  id: Persona;
  label: string;
  tagline: string;
}

// Catalogue des 5 personas — labels FR + taglines de l'ADR-0010 §"Personas v2".
// Les icônes Lucide sont définies côté Svelte (cf. `onboarding/+page.svelte`
// et `parametres/+page.svelte`) pour ne pas importer `@lucide/svelte` ici.
const PERSONA_CATALOG: Record<Persona, PersonaInfo> = {
  student: {
    id: 'student',
    label: 'Étudiant·e / Curieux·se',
    tagline: 'Comprendre votre impact, apprendre les bons réflexes'
  },
  pro_tech: {
    id: 'pro_tech',
    label: 'Professionnel·le tech',
    tagline: 'Optimiser vos prompts, comparer les modèles, exporter pour votre équipe'
  },
  enterprise: {
    id: 'enterprise',
    label: 'Entreprise',
    tagline: 'Piloter votre scope 3 IA, rapport CSRD, forecast budget carbone'
  },
  public_sector: {
    id: 'public_sector',
    label: 'Collectivité / Service public',
    tagline: 'Suivre votre empreinte territoriale, marchés publics frugaux'
  },
  researcher: {
    id: 'researcher',
    label: 'Chercheur·se / Journaliste',
    tagline: 'Reproductibilité, comparaisons inter-modèles, datasets publiables'
  }
};

export function personaLabel(p: Persona): string {
  return PERSONA_CATALOG[p].label;
}

export function personaTagline(p: Persona): string {
  return PERSONA_CATALOG[p].tagline;
}

// ─── Catalogue modules (CDC §4.1) ────────────────────────────────────────

export type ModuleCategory = 'estimation' | 'visualisation' | 'reporting' | 'pedagogie';

export interface ModuleInfo {
  id: ModuleId;
  label: string;
  description: string;
  category: ModuleCategory;
  /** Route SvelteKit du module ; `null` si non encore implémenté. */
  href: string | null;
}

const MODULES: Record<ModuleId, ModuleInfo> = {
  m1: {
    id: 'm1',
    label: 'Estimer un prompt',
    description: 'Mesurer CO₂, énergie, eau, métaux pour une requête unique.',
    category: 'estimation',
    href: '/'
  },
  m2: {
    id: 'm2',
    label: 'Workbench multi-prompts',
    description:
      'Estimer plusieurs prompts en série dans un même atelier. (Différé v1.1+, cf. ADR-0011.)',
    category: 'estimation',
    href: null
  },
  m3: {
    id: 'm3',
    label: 'Comparer modèles',
    description: 'Benchmarker 2 à 8 modèles côte-à-côte sur un même prompt.',
    category: 'estimation',
    href: '/comparer'
  },
  m5: {
    id: 'm5',
    label: 'Exporter rapport',
    description:
      'PDF, CSV ou JSON sourcé — pour intégrer un dossier. (Différé v1.1+, voir M22 Rapport CSRD pour le PDF officiel.)',
    category: 'reporting',
    href: null
  },
  m6: {
    id: 'm6',
    label: 'Géoloc datacenter',
    description: "Préciser l'origine géographique de l'inférence.",
    category: 'visualisation',
    href: null
  },
  m7: {
    id: 'm7',
    label: "Journal d'audit",
    description: 'Ledger chaîné SHA-256 de toutes vos estimations.',
    category: 'reporting',
    href: '/journal'
  },
  m8: {
    id: 'm8',
    label: 'Méthodologie',
    description: 'Comprendre le moteur Monte-Carlo et ses sources.',
    category: 'pedagogie',
    href: '/methodo'
  },
  m9: {
    id: 'm9',
    label: 'Bibliothèque de modèles',
    description: 'Catalogue des 25+ modèles couverts et leurs sources.',
    category: 'pedagogie',
    href: '/modeles'
  },
  m10: {
    id: 'm10',
    label: 'Importer batch',
    description:
      'Charger un CSV/JSON de prompts pour estimation groupée. (Différé v1.1+, backend M18 disponible mais UI dédiée non livrée.)',
    category: 'estimation',
    href: null
  },
  m11: {
    id: 'm11',
    label: 'Extension navigateur',
    description: 'Capter vos vrais usages ChatGPT/Claude depuis votre browser.',
    category: 'estimation',
    href: null
  },
  m12: {
    id: 'm12',
    label: 'Datacenters Europe',
    description: 'Carte interactive des sites avec drill-down énergie.',
    category: 'visualisation',
    href: '/datacenters'
  },
  m13: {
    id: 'm13',
    label: 'Simulateur « Et si...? »',
    description: '7 leviers temps réel pour explorer des scénarios.',
    category: 'estimation',
    href: '/simuler'
  },
  m14: {
    id: 'm14',
    label: 'À propos / Crédits',
    description: 'Sources, licences, contributeurs, méthodologie courte.',
    category: 'pedagogie',
    href: null
  },
  m15: {
    id: 'm15',
    label: 'Tableau de bord personnel',
    description: 'Vos usages agrégés (jour, semaine, mois).',
    category: 'visualisation',
    href: '/suivi'
  },
  m16: {
    id: 'm16',
    label: 'Forecast 12 mois',
    description: "Projection annuelle avec bande d'incertitude et leviers.",
    category: 'reporting',
    href: null
  },
  m17: {
    id: 'm17',
    label: 'Datasheet scientifique',
    description: 'Bilan complet pour reproductibilité scientifique.',
    category: 'reporting',
    href: '/datasheets'
  },
  m18: {
    id: 'm18',
    label: 'Batch CSV → rapport',
    description: 'Import + agrégation pour un rapport sourcé prêt à partager.',
    category: 'reporting',
    href: null
  },
  m19: {
    id: 'm19',
    label: 'Équipe multi-utilisateurs',
    description: 'RBAC léger pour partager une installation entre collègues.',
    category: 'reporting',
    href: null
  },
  m20: {
    id: 'm20',
    label: 'Territoire France',
    description: 'Cartographie RTE IRIS + Sankey énergétique territorial.',
    category: 'visualisation',
    href: '/territoire'
  },
  m21: {
    id: 'm21',
    label: 'Alertes & seuils',
    description: 'Notifications locales quand un seuil personnel est franchi.',
    category: 'reporting',
    href: null
  },
  m22: {
    id: 'm22',
    label: 'Rapport réglementaire (CSRD/AGEC)',
    description: 'PDF signé + bundle PROV-O conforme aux normes.',
    category: 'reporting',
    href: '/rapport-csrd'
  },
  m23: {
    id: 'm23',
    label: 'Marchés publics IA frugale',
    description: "Modèles de cahiers des charges pour appels d'offre.",
    category: 'reporting',
    href: null
  },
  m24: {
    id: 'm24',
    label: 'Apprendre',
    description: 'Mini-cours, bonnes pratiques, parcours guidés.',
    category: 'pedagogie',
    href: null
  },
  m25: {
    id: 'm25',
    label: 'Objectifs & habitudes',
    description: 'Eco-budget personnel hebdomadaire avec encouragements.',
    category: 'pedagogie',
    href: null
  }
};

export function moduleLabel(id: ModuleId): string {
  return MODULES[id].label;
}

export function moduleDescription(id: ModuleId): string {
  return MODULES[id].description;
}

export function moduleHref(id: ModuleId): string | null {
  return MODULES[id].href;
}

export function moduleCategory(id: ModuleId): ModuleCategory {
  return MODULES[id].category;
}

export function moduleInfo(id: ModuleId): ModuleInfo {
  return MODULES[id];
}

// ─── C32.2 — Pourquoi ce module est dans le bundle de ce persona ? ──────
// Tooltip pédagogique affiché au survol des modules pré-cochés dans
// l'onboarding (étape 4 Bundle) et dans /parametres. Le texte est
// persona-spécifique : un même module M1 sera justifié différemment pour
// un étudiant (« mesurer ton premier prompt en grammes ») et pour une
// entreprise (« référence auditable pour ton scope 3 IA »).

const MODULE_PERSONA_REASON: Record<Persona, Partial<Record<ModuleId, string>>> = {
  student: {
    m1: "Mesurer chaque prompt pour comprendre l'ordre de grandeur (CO₂, eau).",
    m8: 'Voir la méthodologie de mesure expliquée simplement.',
    m13: "Explorer 'et si je raccourcis mes prompts ?' ou 'et si je change de modèle ?'.",
    m15: 'Suivre ton usage semaine par semaine, équivalences humaines incluses.',
    m25: 'Fixer un objectif mensuel + alerte quand tu le dépasses.'
  },
  pro_tech: {
    m1: 'Mesurer chaque appel API individuellement (ton terrain de jeu).',
    m3: 'Comparer 3 modèles côte-à-côte pour choisir le plus frugal.',
    m7: 'Ledger SHA-256 chaîné pour le reporting interne / audit.',
    m8: 'Voir les formules par méthodologie (AFNOR + EcoLogits).',
    m9: 'Catalogue 25+ modèles avec P5/P50/P95 + vendor disclosure (Mistral × ADEME, Google, Meta).',
    m13: "Explorer les leviers d'optimisation avant de modifier ton intégration."
  },
  enterprise: {
    m1: "Référence d'estimation auditable pour ton scope 3 IA.",
    m7: 'Ledger chaîné = preuve non-altération pour audits CSRD.',
    m12: 'Comprendre où tournent vos LLM en Europe (PUE + mix élec).',
    m15: 'Vue agrégée de votre usage (admin / RSE).',
    m17: 'Datasheet Gebru pour reproductibilité scientifique.',
    m20: 'Empreinte territoriale FR par IRIS (RTE × ComparIA).',
    m22: 'Rapport CSRD/AGEC trimestriel signé + PROV-O.',
    m25: 'Définir un plafond carbone mensuel avec alerte webhook ou email.'
  },
  public_sector: {
    m1: 'Mesurer chaque prompt avec rigueur méthodologique sourcée Etalab 2.0.',
    m8: 'Méthodologie AFNOR SPEC 2314 + sources officielles FR (ADEME, RTE).',
    m12: 'Cartographier les datacenters européens (carte Leaflet + drill-down).',
    m17: 'Datasheet scientifique réutilisable comme template marchés publics.',
    m20: 'Empreinte par IRIS RTE : différenciateur FR unique de Sobr.ia.',
    m22: 'Rapport réglementaire AGEC PDF + JSON-LD PROV-O.'
  },
  researcher: {
    m1: 'Atelier reproductible (seed SOBRIA_SEED=42).',
    m3: 'Benchmark inter-modèles multi-méthodologie (AFNOR + EcoLogits).',
    m7: 'Ledger chaîné SHA-256 = preuve non-altération pour reviewers.',
    m8: 'Méthodologie complète + sources DOI (EcoLogits, AFNOR SPEC 2314).',
    m9: 'Catalogue P5/P50/P95 + vendor disclosure (transparence multi-méthodo).',
    m17: 'Datasheet Gebru 2018 auto-générée + JSON-LD PROV-O.'
  }
};

/**
 * Renvoie l'explication « pourquoi ce module est dans ton bundle persona »
 * ou `undefined` si pas de raison spécifique (module hors bundle par
 * défaut ou persona null).
 *
 * Utilisé comme attribut `title` natif sur les module-rows.
 */
export function moduleReason(persona: Persona | null, id: ModuleId): string | undefined {
  if (persona === null) return undefined;
  return MODULE_PERSONA_REASON[persona][id];
}

export const CATEGORY_LABELS: Record<ModuleCategory, string> = {
  estimation: 'Estimation',
  visualisation: 'Visualisation',
  reporting: 'Reporting',
  pedagogie: 'Pédagogie'
};

/** Helper de garde de route : true si le module est dans le bundle actif. */
export function isModuleEnabled(id: ModuleId): boolean {
  return get(preferences).enabled_modules.includes(id);
}
